core/src/membank_controller_io.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include "nes/core/imembank.h"
+#include "nes/core/ines_controller.h"
+
+namespace n_e_s::core {
+
+class MemBankControllerIO : public IMemBank {
+public:
+    MemBankControllerIO(uint8_t slot, INesController *controller)
+            : controller_(controller),
+              controller_addr_{
+                      slot == 0 ? uint16_t(0x4016) : uint16_t(0x4017)} {}
+
+    bool is_address_in_range(uint16_t addr) const override {
+        return addr == controller_addr_;
+    }
+
+    uint8_t read_byte(uint16_t addr) const override {
+        if (addr == controller_addr_) {
+            // Source: https://wiki.nesdev.com/w/index.php/Standard_controller
+            // In the NES and Famicom, the top three (or five) bits are not
+            // driven, and so retain the bits of the previous byte on the bus.
+            // Usually this is the most significant byte of the address of the
+            // controller port—0x40. Certain games (such as Paperboy) rely on
+            // this behavior and require that reads from the controller ports
+            // return exactly $40 or $41 as appropriate.
+            uint8_t data = 0x40;
+            uint8_t btn = 0x00;
+
+            switch (read_cnt_++) {
+            case 0:
+                btn = controller_->get(INesController::Button::A) ? 0x01 : 0x00;
+                break;
+            case 1:
+                btn = controller_->get(INesController::Button::B) ? 0x01 : 0x00;
+                break;
+            case 2:
+                btn = controller_->get(INesController::Button::Select) ? 0x01
+                                                                       : 0x00;
+                break;
+            case 3:
+                btn = controller_->get(INesController::Button::Start) ? 0x01
+                                                                      : 0x00;
+                break;
+            case 4:
+                btn = controller_->get(INesController::Button::Up) ? 0x01
+                                                                   : 0x00;
+                break;
+            case 5:
+                btn = controller_->get(INesController::Button::Down) ? 0x01
+                                                                     : 0x00;
+                break;
+            case 6:
+                btn = controller_->get(INesController::Button::Left) ? 0x01
+                                                                     : 0x00;
+                break;
+            case 7:
+                btn = controller_->get(INesController::Button::Right) ? 0x01
+                                                                      : 0x00;
+                break;
+            }
+            data = static_cast<uint8_t>(data & static_cast<uint8_t>(0xFE)) |
+                   btn;
+            return data;
+        }
+        return 0x00;
+    }
+
+    void write_byte(uint16_t addr, uint8_t data) override {
+        if (addr == controller_addr_) {
```

Comment:
From https://wiki.nesdev.com/w/index.php/Standard_controller "Though both are polled from a write to $4016, controller 1 is read through $4016, and controller 2 is separately read through $4017."
So seems like controller 2 needs to use 0x4016 here

Reply:
Fixed - now using only one membank controller for both controllers to handle 'strobe' on address $4016 for both controllers.